File: ios/Runner/Modules/Channel/Model/FlutterChannelMethod.swift
//
//  FlutterChannelMethod.swift
//  Runner
//
//  Created by Manu Espeso on 12/9/25.
//

enum FlutterChannelMethod {
    /// Flutter -> Native
    case storeAppDownloaded
    case listenShakeAction
    case retrieveDeviceDecimalSeparator
    /// Native -> Flutter
    case onShakeDevice
    
    fileprivate enum FlutterChannelMethodNaming: String {
        case isAppDownloadedFromStore, listenShakeDeviceAction, getDeviceDecimalSeparator, onShakeDevice
    }
    
    init?(_ rawValue: String, arguments: [String: Any]?) {
        let value: FlutterChannelMethod? = switch rawValue {
            case FlutterChannelMethodNaming.isAppDownloadedFromStore.rawValue:
                    .storeAppDownloaded
            case FlutterChannelMethodNaming.listenShakeDeviceAction.rawValue:
                    .listenShakeAction
            case FlutterChannelMethodNaming.getDeviceDecimalSeparator.rawValue:
                    .retrieveDeviceDecimalSeparator
            case FlutterChannelMethodNaming.onShakeDevice.rawValue:
                    .onShakeDevice
            default:
                nil
        }
        
        guard let value else { return nil }
        
        self = value
    }
    
    var name: String {
        switch self {
            case .storeAppDownloaded:
                return FlutterChannelMethodNaming.isAppDownloadedFromStore.rawValue
            case .listenShakeAction:
                return FlutterChannelMethodNaming.listenShakeDeviceAction.rawValue
            case .retrieveDeviceDecimalSeparator:
                return FlutterChannelMethodNaming.getDeviceDecimalSeparator.rawValue
            case .onShakeDevice:
                return FlutterChannelMethodNaming.onShakeDevice.rawValue
        }
    }
}

File: android/app/src/main/kotlin/com/labhouse/app/modules/channel/model/FlutterChannelMethod.kt
package com.labhouse.app.modules.channel.model

sealed class FlutterChannelMethod {
    data object StoreAppDownloaded : FlutterChannelMethod()
    data object ListenShakeAction : FlutterChannelMethod()
    data object RetrieveDeviceDecimalSeparator : FlutterChannelMethod()
    data object OnShakeDevice : FlutterChannelMethod()
    // data class NewCase(val payload: Map<String, Any?>) : FlutterChannelMethod()

    private object Naming {
        const val IS_APP_DOWNLOADED_FROM_STORE = "isAppDownloadedFromStore"
        const val LISTEN_SHAKE_DEVICE_ACTION = "listenShakeDeviceAction"
        const val GET_DEVICE_DECIMAL_SEPARATOR = "getDeviceDecimalSeparator"
        const val ON_SHAKE_DEVICE = "onShakeDevice"
        // const val newCase = "newCase"
    }

    companion object {
        fun from(raw: String, arguments: Any?): FlutterChannelMethod? = when (raw) {
            Naming.IS_APP_DOWNLOADED_FROM_STORE -> StoreAppDownloaded
            Naming.LISTEN_SHAKE_DEVICE_ACTION -> ListenShakeAction
            Naming.GET_DEVICE_DECIMAL_SEPARATOR -> RetrieveDeviceDecimalSeparator
            Naming.ON_SHAKE_DEVICE -> OnShakeDevice
            // Naming.newCase -> NewCase(arguments as? Map<String, Any?> ?: emptyMap())
            else -> null
        }
    }

    val name: String
        get() = when (this) {
            StoreAppDownloaded -> Naming.IS_APP_DOWNLOADED_FROM_STORE
            ListenShakeAction -> Naming.LISTEN_SHAKE_DEVICE_ACTION
            RetrieveDeviceDecimalSeparator -> Naming.GET_DEVICE_DECIMAL_SEPARATOR
            OnShakeDevice -> Naming.ON_SHAKE_DEVICE
            // is NewCase -> Naming.newCase
        }
}
